libethereum/State.cpp

```diff
@@ -427,39 +426,45 @@ h256 State::storageRoot(Address const& _id) const
 	return EmptyTrie;
 }
 
-bytes const& State::code(Address const& _contract) const
+bytes const& State::code(Address const& _a) const
 {
-	if (!addressHasCode(_contract))
+	if (!addressHasCode(_a))
 		return NullBytes;
-	ensureCached(_contract, true, false);
-	return m_cache[_contract].code();
+	return account(_a, true)->code();
 }
 
-h256 State::codeHash(Address const& _contract) const
+h256 State::codeHash(Address const& _a) const
 {
-	if (!addressHasCode(_contract))
+	if (Account const* a = account(_a))
+	{
+		if (a->isFreshCode())
+			return sha3(a->code());
+		else
+			return a->codeHash();
+	}
+	else
 		return EmptySHA3;
-	if (m_cache[_contract].isFreshCode())
-		return sha3(code(_contract));
-	return m_cache[_contract].codeHash();
 }
 
-size_t State::codeSize(Address const& _contract) const
+size_t State::codeSize(Address const& _a) const
 {
-	if (!addressHasCode(_contract))
-		return 0;
-	if (m_cache[_contract].isFreshCode())
-		return code(_contract).size();
-	auto& codeSizeCache = CodeSizeCache::instance();
-	h256 codeHash = m_cache[_contract].codeHash();
-	if (codeSizeCache.contains(codeHash))
-		return codeSizeCache.get(codeHash);
-	else
+	if (Account const* a = account(_a))
 	{
-		size_t size = code(_contract).size();
-		codeSizeCache.store(codeHash, size);
-		return size;
+		if (a->isFreshCode())
+			return code(_a).size();
+		auto& codeSizeCache = CodeSizeCache::instance();
```

Comment:
Can you please elaborate? It has to be global in order to speed up code size lookup for two contracts that have the same code - that is its only purpose. In addition, we can add another cache inside the account, but I'm not sure if that is an improvement because the account data structure will grow.